patterns.md

```diff
@@ -0,0 +1,62 @@
+except for TODO where the constant is treated as if it was inlined as a pattern (and the usual `match` tree is constructed).
+
+## Soundness concerns
+
+Most of the time there are no extra soundness concerns due to const-patterns; except for surprising behavior nothing can go wrong when the structural equality check gives a wrong answer.
+However, there is one exception: some constants participate in exhaustiveness checking.
+If a pattern is incorrectly considered exhaustive, that leads to a critical soundness bug.
+
+Exhaustiveness checking is done for constants of type TODO, as well as `&[u8]` and `&str` (but no other types with references).
+This means we can write:
+```rust
+#![feature(exclusive_range_pattern)]
+#![feature(half_open_range_patterns)]
+const PAT: &[u8] = &[0];
+
+pub fn test(x: &[u8]) -> bool {
+    match x {
+        PAT => true,
+        &[] => false,
+        &[1..] => false,
+        &[_, _, ..] => false
+    }
+}
+```
+
+To ensure soundness of exhaustiveness checking, it is crucial that all data considered this check is fully immutable.
+In particular, for constants of reference type, it is important that they only point to immutable data.
+For this reason, the static const checks reject references to `static` items.
+This is a new soundness concern that otherwise does not come up during CTFE (where *reading from* a mutable `static` during const initialization is a problem, but just referencing a mutable `static` is not).
+A more precise check could be possible, but is non-trivial: even an immutable `static` could point to a mutable `static`; that would have to be excluded.
+(We could, in principle, also rely on it being UB to have a shared reference to mutable memory, but for now we prefer not to rely on the aliasing model like that---aliasing of global pointers is a tricky subject.)
+
+*Dynamic check.*
+To dynamically ensure constants only point to read-only data, we maintain a global invariant:
+pointers to a `static` (including memory that was created as part of interning a `static`) may never "leak" to a non-`static`.
+All memory outside of `static`s is marked as immutable.
+As a consequence, non-`static`s recursively only point to immutable memory.
+
+This check is uncomfortably close to the static checks, and fragile due to its reliance on a global invariant.
+Longer-term, it would be better to move to a more local check, by making validation check if consts are recursively read-only.
+This check (unlike our current validation) would have to descend through pointers to `static`.
+
+An alternative could be to wait for the [value tree proposal](https://github.com/rust-lang/compiler-team/issues/323) to be implemented.
+Then we could check during value tree construction that all memory is read-only (if we use the CTFE machine in `const`-mode, that will happen implicitly), and thus we know for sure we can rely on the value tree to be sound to use for exhaustiveness checking.
```

Comment:
@oli-obk the more I think about it, the more I like this: as far as the dynamic checks are concerned, constants can point to mutable data just fine as long as we never construct a value tree for them (or as long as the mutable part is in the "fallback" part of the value tree that we might have to support "bad" patterns that do not fit a value tree, like raw pointers, function pointers, floats).

Reply:
The problem isn't just a value tree pointing to a mutable allocation, it's a mutable allocation being used during the const evaluation of the initializer of a constant that will be converted to a value tree. It's not really unsound and I don't know of any problems, we just have had the policy of "running the initializers of a constant manually (copy paste) gives the same result as reading the constant", and this would break it.

Reply:
That is an entirely orthogonal concern (not even specific to patterns) that we [already cover here](https://github.com/rust-lang/const-eval/blob/master/const.md#reading-statics). But I think you are mixing up *initializing a const reads from a static that could later be mutated* and *the final value of a const (transitively) points to a mutable static*. These are two very different and orthogonal concerns, in my view.

Reply:
... I had something written and then realized we are only talking about dynamic checks... Yea, it's all fine for dynamic checks. I just can't come up with static checks that are more relaxed than today's.

Reply:
Right, I see static checks as something we can only really start talking about once we have established what the dynamic property is that we want to approximate.